lib/Sema/CSDiagnostics.cpp

```diff
@@ -1061,15 +1061,30 @@ bool MissingExplicitConversionFailure::diagnoseAsError() {
     anchor = paren->getSubExpr();
 
   auto fromType = getFromType();
-  Type toType = getToType();
+  auto toType = getToType();
 
   if (!toType->hasTypeRepr())
     return false;
 
   bool useAs = TC.isExplicitlyConvertibleTo(fromType, toType, DC);
   bool useAsBang = !useAs && TC.checkedCastMaySucceed(fromType, toType, DC);
-  if (!useAs && !useAsBang)
+  bool isNonContextualExplicitCoercion =
+      getLocator()->isLastElement(
+          ConstraintLocator::PathElementKind::ExplicitTypeCoercion) &&
+      getContextualTypePurpose() == CTP_Unused;
+  if (!useAs && !useAsBang) {
+    if (isNonContextualExplicitCoercion) {
+      // If the coercion involves two bounded generic types defer this
+      // to repair the failure with a GenericArgumentsMismatch fix.
+      if (fromType->is<BoundGenericType>() && toType->is<BoundGenericType>())
```

Comment:
This is an indication of the solver's failure to record a correct fix for the situation which I'm guessing is something like this:

```swift
struct S<T> {}

var s: S<Int> = S()
_ = s as S<String>
```

Would you mind fixing that in the separate PR (to make sure that we record `GenericArgumentsMismatch` before we merge this one? This way we wouldn't need to touch this code at all here.

Reply:
Sure @xedin  I'll put that in another PR. Thank's for all the reviews here :))